### ISSUE TITLE
FunctionsStartup is not being called, causing function dependency resolution to fail.

### ISSUE THREAD

**Issue by worleywa7** [Azure/azure-functions-dotnet-extensions#5]
We downloaded the sample project, put a break point on             builder.Services.AddScoped<IGreeter, SampleGreeter>(); expecting it to be called when the runtime started and discovered functions. Unfortunately it did no, so we attempted to invoke the function directly to see if the runtime would call the FunctionStartup class upon first function invocation. Unfortunately it did not and both times resulted in:

Executed 'SampleFunction' (Failed, Id=eeaabc43-d8a6-4815-9fe2-0834f7d4c47b)
[5/8/2019 9:58:22 PM] Microsoft.Extensions.DependencyInjection.Abstractions: Unable to resolve service for type 'Microsoft.Azure.Functions.Samples.DependencyInjectionBasic.IGreeter' while attempting to activate 'Microsoft.Azure.Functions.Samples.DependencyInjectionBasic.SampleFunction'.

Is there something we are missing? It does not appear that the FunctionsStartup method is actually being called by the runtime.

**Comment by bentmar** [comment]
https://github.com/Azure/azure-functions-dotnet-extensions/pull/8
heres the fix

**Comment by worleywa7** [comment]
Thank you very much for your time, we got it working!

**Comment by bentmar** [comment]
@worleywa7  Just curious, have you tried publishing a function with DI to azure? 

Mine works locally, but crashes instantly in the cloud with a DependencyIncetion error..

**Comment by bentmar** [comment]
Okay, let the randomness begin..

1.0.27 and 1.0.28 doesnt work in azure for dependency resolution, BUT 1.026 does. 

And when downgradig to 1.0.26 it suddenly works locally aswell!

**Comment by worleywa7** [comment]
Looks like we're getting the same issue as well; wasn't readily apparent until now. We're going to downgrade as well; see what we come up with.

**Comment by worleywa7** [comment]
I downgraded back to 1.0.26 and things returned to throwing the DI errors...I'm going to try 1.0.28.

**Comment by worleywa7** [comment]
Looks like we're back to the drawing board: 

'The function runtime is unable to start. Towar.Cloud.Functions: Method not found: 'Microsoft.Extensions.DependencyInjection.IServiceCollection Microsoft.Azure.Functions.Extensions.DependencyInjection.IFunctionsHostBuilder.get_Services()'.
Session Id: 37b85485f2624771a77400e47d07ef45

Timestamp: 2019-05-16T15:13:38.648Z'

**Comment by worleywa7** [comment]
So we did what you said at first, and it worked; that's why I closed the ticket.. I saw exactly what I wanted to see in my project. However, since then, I added nugets from two .NET Standard libraries, injected a dependency from one of those libraries, and have created functions that rely on this DIs.

**Comment by antichaosdb** [comment]
I tried the scopes sample - no other references.  As provided with 1.0.26 it failed locally and on Azure.  With Dependencies updated (1.0.28) it runs locally but when deployed to Azure I get :
Unable to resolve service for type 'Microsoft.Azure.Functions.Samples.DependencyInjectionScopes.MyServiceA' while attempting to activate 'Microsoft.Azure.Functions.Samples.DependencyInjectionScopes.SampleFunction'.

Is there a way to do some logging in the startup function so we can see if it being run when deployed?

**Comment by worleywa7** [comment]
So what it ended up being was, for me at least, there was a dependency that I forgot to inject. There was a single "Object cannot be null" line in a log on the azure portal that seemed to be the issue. We're back to running our functions on 1.0.28 in the Cloud.

**Comment by jedjohan** [comment]
The only way to get DI to work for me when deployed to Azure is using 1.0.26. 

1.0.26, 1.0.27, 1.0.28 all works fine local

**Comment by eliashdezr** [comment]
`1.0.26` doesn't work locally for me, I'm testing it with the sample project.

**Comment by antichaosdb** [comment]
Finally got this working.  According to [(https://github.com/Azure/azure-functions-host/issues/3386#issuecomment-497492442)]  the suggestion to add a reference to ExtensionsMetadataGenerator is bad.  Remove this reference if you have it.

**Comment by ErikAndreas** [comment]
1.0.29 not working, same issue as described in https://github.com/Azure/azure-functions-dotnet-extensions/issues/14 (closed due to author moving on to Aws Lambda...)

Microsoft.Extensions.DependencyInjection.Abstractions: Unable to resolve service for type '...' while attempting to activate '...'

**Comment by itorian** [comment]
That's totally fine for me on development and production. Here's project configurations:

```
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v2</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.29" />
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
  </ItemGroup>
```

Notice i'm using, function v2, netcoreapp3.0, Sdk.Function 1.0.29 and Function.Extension 1.0.0.

Sample project here https://github.com/itorian/AzureFunctionDISample

**Comment by robertracz** [comment]
Same here... Happening when I reference 

`<PackageReference Include="Microsoft.Extensions.Http" Version="3.0.0" />`

**Comment by dacris** [comment]
I had the same issue today. I removed some Nuget packages and added a namespace declaration to my Startup.cs and the problem was fixed. Using SDK 1.0.29. Not sure if this matters, but check if your Startup class has a namespace.

**Comment by apavelm** [comment]
At this moment the solution is only to downgrade to .NET Core 2.2

**Comment by eliashdezr** [comment]
I'm on the same boat, these combinations of NuGet versions are not working:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
    <PackageReference Include="Microsoft.Extensions.Http" Version="3.0.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.0-preview2" />
    <PackageReference Include="Utf8Json" Version="1.3.7" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```
Error is
` Microsoft.Extensions.DependencyInjection.Abstractions: Unable to resolve service for type`

**Comment by arkiaconsulting** [comment]
```
<PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.0" />
```
is out.

Still
```
Microsoft.Azure.Functions.Extensions: Method not found: 'Microsoft.Extensions.DependencyInjection.IServiceCollection Microsoft.Azure.WebJobs.IWebJobsBuilder.get_Services()'.
Value cannot be null. (Parameter 'provider')
```

**Comment by eliashdezr** [comment]
Is the runtime version still on preview?

https://docs.microsoft.com/en-us/azure/azure-functions/functions-versions

**Comment by arkiaconsulting** [comment]
Tested locally with v3 instead of v3-preview with no success (No function runtime available)

**Comment by eliashdezr** [comment]
And this error still persist:
´Microsoft.Extensions.DependencyInjection.Abstractions: Unable to resolve service for type ´

**Comment by coolhome** [comment]
I had issues when using preview without install visual studio 16.4 preview. Try installing visual studio 16.4 with the official release?

**Comment by eliashdezr** [comment]
I'm still getting the following error while trying to do a `GET` on a `HttpTrigger` function: 

` Microsoft.Extensions.DependencyInjection.Abstractions: Unable to resolve service for type`

This is my current setup:
 - `Visual Studio 16.5.0 Preview 1`
 - `azure-functions-core-tools@3.0.1740`
 - `npm@v13.2.0`

And this is my `.csproj`
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
    <PackageReference Include="Microsoft.Extensions.Http" Version="3.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.1" />
    <PackageReference Include="Utf8Json" Version="1.3.7" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

`Startup.cs` file:
```csharp
using Microsoft.Azure.Functions.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection;
using Demo.Domain.Interfaces.Repository;

namespace Demo.Functions
{
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddSingleton<IDemoRepo, DemoRepo>();
        }
    }
}
```

`local.settings.json`
```json
{
  "IsEncrypted": false,
  "Values": {
    "AzureWebJobsStorage": "UseDevelopmentStorage=true",
    "FUNCTIONS_WORKER_RUNTIME": "dotnet",
    "FUNCTIONS_EXTENSION_VERSION": "~3"
  }
}
```

`host.json`
```json
{
  "version": "2.0",
  "healthMonitor": {
    "enabled": true,
    "healthCheckInterval": "00:00:10",
    "healthCheckWindow": "00:02:00",
    "healthCheckThreshold": 6,
    "counterThreshold": 0.80
  },
  "functionTimeout": "00:10:00"
}
```

**Comment by wtt235-bh** [comment]
This seems like it could also be related to https://github.com/Azure/azure-functions-host/issues/5401.

**Comment by fabiocav** [comment]
Closing this issue as there is a mix of different problems tracked here, and they should be resolved with the release of Azure Functions 3.0 (many of the issues here are related to running on Azure Functions 2.0 while targeting .NET Core 3.0).

If you continue to have issues, please open a new issue with details and we'll investigate them. Thank you!
